Fix unsuccessful KubernetesPod final_state call when `is_delete_operator_pod=True`

MatthewRBruce opened the pull request:
If a Kubernetes Pod ends in a state other than `SUCCESS` and `is_delete_operator_pod` is True, then use the `final_state` from the previous `create_new_pod_for_operator` call since the pod is already deleted and the current state can't be re-read.

closes: https://github.com/apache/airflow/issues/15456

MatthewRBruce committed:
Fix unsuccessful KubernetesPod final_state call

If a Kubernetes Pod ends in a state other than SUCCESS and
is_delete_operator_pod is True, then use the `final_state` from the
`create_new_pod_for_operator` call since the pod is already deleted and
the current state can't be re-read.

ephraimbuddy reviewed (changes requested):
@MatthewRBruce Thanks for this PR. Can you add a test to avoid regression

jedcunningham commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -365,7 +365,10 @@ def execute(self, context) -> Optional[str]:
             if final_state != State.SUCCESS:
-                status = self.client.read_namespaced_pod(self.pod.metadata.name, self.namespace)
+                if self.is_delete_operator_pod:
+                    status = final_state
```
I wonder if we should make this actually dump the whole pod like it would if `is_delete_operator_pod` was false, instead of the state. It would take a little refactoring to make it happen, but seems like it could be pretty helpful in that scenario?

Maybe this exception should always just include the state, then have `create_new_pod_for_operator` log the full pod on failure?

MatthewRBruce replied:
That's a good idea - we could modify `monitor_pod` to return a Tuple of `[State, event, result]` instead of just `[State, result]` (https://github.com/apache/airflow/blob/b844621a6f42cc20b3103bc70c774023268f8de8/airflow/providers/cncf/kubernetes/utils/pod_launcher.py#L170).  Then `create_new_pod_for_operator` could pass that back through to `execute` too (or `create_new_pod_for_operator` could read the pod again in it's `finally` clause and pass that back)

MatthewRBruce replied:
Okay, I've updated the code to return the full V1Pod back through from `monitor_pod`.  This allows Airflow to log the full pod description in the event of a failure regardless of `is_delete_pod_operator`. 

I modified the original test that checked whether `read_namespaced_pod` was called in the event of a failure, since it doesn't need to be called either way now - it just checks that the failed pod description is reported in the Exception.

MatthewRBruce committed:
monitor_pod always return the full pod state

The monitor_pod method now always returns the full V1Pod object returned
from read_namespaced_pod now.  This allows the
KubernetesPodOperator.execute method output the full pod description on
a failure regardless of is_delete_pod_operator.

MatthewRBruce committed:
Fix tests in kubernetes_tests

jedcunningham commented on airflow/providers/cncf/kubernetes/utils/pod_launcher.py:
```diff
@@ -129,7 +129,7 @@ def start_pod(self, pod: V1Pod, startup_timeout: int = 120):
-    def monitor_pod(self, pod: V1Pod, get_logs: bool) -> Tuple[State, Optional[str]]:
+    def monitor_pod(self, pod: V1Pod, get_logs: bool) -> Tuple[State, V1Pod, Optional[str]]:
         """
         Monitors a pod and returns the final state
```
```suggestion
        Monitors a pod and returns the final state, pod, and xcom result
```

jedcunningham commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -358,15 +358,14 @@ def execute(self, context) -> Optional[str]:
             if len(pod_list.items) == 1:
                 try_numbers_match = self._try_numbers_match(context, pod_list.items[0])
-                final_state, result = self.handle_pod_overlap(
+                final_state, pod_info, result = self.handle_pod_overlap(
```
```suggestion
                final_state, _, result = self.handle_pod_overlap(
```
Since we don't use it?

MatthewRBruce replied:
It's used on line 368 if the resulting pod isn't successful, no?

jedcunningham replied:
Oh yeah, missed that!

jedcunningham commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -358,15 +358,14 @@ def execute(self, context) -> Optional[str]:
             else:
                 self.log.info("creating pod with labels %s and launcher %s", labels, launcher)
-                final_state, _, result = self.create_new_pod_for_operator(labels, launcher)
+                final_state, _, pod_info, result = self.create_new_pod_for_operator(labels, launcher)
```
```suggestion
                final_state, _, pod, result = self.create_new_pod_for_operator(labels, launcher)
```

I think I like `pod` better than `pod_info`? Only doing this comment once, but applies everywhere.

MatthewRBruce replied:
👍

jedcunningham commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -518,17 +517,17 @@ def create_new_pod_for_operator(self, labels, launcher) -> Tuple[State, k8s.V1Po
         except AirflowException:
             if self.log_events_on_failure:
-                for event in launcher.read_pod_events(self.pod).items:
+                for event in pod_info.items:
```
```suggestion
                for event in launcher.read_pod_events(self.pod).items:
```

I think we want to leave this as-is. We wont get a `pod(_info)` back, plus `read_pod_events` just needs the namespace and name, which we can always use `self.pod` for.

jedcunningham commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -518,17 +517,17 @@ def create_new_pod_for_operator(self, labels, launcher) -> Tuple[State, k8s.V1Po
                 self.log.debug("Deleting pod for task %s", self.task_id)
                 launcher.delete_pod(self.pod)
-        return final_state, self.pod, result
+        return final_state, self.pod, pod_info, result
```
Returning the pod twice?
```suggestion
        return final_state, pod_info, result
```

jedcunningham commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -545,17 +544,17 @@ def monitor_launched_pod(self, launcher, pod) -> Tuple[State, Optional[str]]:
         if final_state != State.SUCCESS:
             if self.log_events_on_failure:
-                for event in launcher.read_pod_events(pod).items:
+                for event in pod_info.items:
```
```suggestion
                for event in launcher.read_pod_events(pod).items:
```

MatthewRBruce replied:
The problem here is that we've already deleted the pod on like 550, so we can't call `read_pod_events` against it

jedcunningham replied:
The pod events are still available after deleting the pod. We can't use `read_namespaced_pod`, but `read_pod_events` should still work. We also won't get `pod_info` back here in all cases.

MatthewRBruce replied:
Ah, that makes sense then.  I was just going over the code and your other comments and I wonder if it wouldn't be easier and better to re-read the pod like `self.pod = launcher.read_pod(self.pod)` in the `finally` blocks of `monitor_launched_pod` and `create_new_pod_for_operator`:
https://github.com/apache/airflow/blob/main/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py#L528
https://github.com/apache/airflow/blob/main/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py#L550

and then just log `self.pod` here:
https://github.com/apache/airflow/blob/main/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py#L369

Way fewer changes and a lot less passing the pod around through the various functions.

jedcunningham replied:
That works for me, just make sure any exceptions when reading it again don't bubble up.

MatthewRBruce replied:
I was poking at this last night and I didn't end up being overly keen on overwriting the original pod spec in `self.pod` with the remote pod after completion.  I went back to the original idea in the PR, but I renamed all the `pod_info` to `remote_pod` - going with the `pod` variable name as you suggested ended up conflating the returned pod with the `pod` parameter passed into some functions and `self.pod` which just seemed confusing.

If you wouldn't mind taking another look at the PR it would be much appreciated :)

jedcunningham replied:
Yeah, I went back and forth on that yesterday as well. There is precedence for setting `self.pod` to be the remote pod, but I think its easier to grok as it is now.

Thanks for being receptive to iterating on this 🍺

MatthewRBruce replied:
> Thanks for being receptive to iterating on this

No problem I think we all want a solution that everyone likes 😄

MatthewRBruce committed:
Update returned pod to be called remote_pod

jedcunningham commented on airflow/providers/cncf/kubernetes/utils/pod_launcher.py:
```diff
@@ -170,7 +170,8 @@ def monitor_pod(self, pod: V1Pod, get_logs: bool) -> Tuple[State, Optional[str]]
             time.sleep(2)
-        return self._task_status(self.read_pod(pod)), result
+        pod_info = self.read_pod(pod)
+        return self._task_status(pod_info), pod_info, result
```
```suggestion
        remote_pod = self.read_pod(pod)
        return self._task_status(remote_pod), remote_pod, result
```

jedcunningham commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -545,17 +544,17 @@ def monitor_launched_pod(self, launcher, pod) -> Tuple[State, Optional[str]]:
         if final_state != State.SUCCESS:
             if self.log_events_on_failure:
-                for event in launcher.read_pod_events(pod).items:
+                for event in launcher.read_pod_events(self.pod).items:
```
```suggestion
                for event in launcher.read_pod_events(pod).items:
```
Since we take a pod as an arg, lets use that instead just so it behaves as one would expect, even though right now its the same thing.

MatthewRBruce replied:
Works for me.  Do we want to update the one on line 555 to match too then? It's `self.pod` in main

jedcunningham replied:
Yeah, good catch, missed that one.

MatthewRBruce replied:
👍 Fixed up.  Hopefully no weird CI failures this time.

jedcunningham reviewed (approved):
Other than the `self.pod`->`pod`, LGTM!
